sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala

```diff
@@ -1164,4 +1166,30 @@ object DateTimeUtils {
     threadLocalTimestampFormat.remove()
     threadLocalDateFormat.remove()
   }
+
+  class MicrosCalendar(tz: TimeZone) extends GregorianCalendar(tz, Locale.US) {
+    def getMicros(): SQLTimestamp = {
```

Comment:
Can we have some comments to explain the behavior? Seems to me it's
1. for `.1`, it's 1000 microseconds
2. for `.1234`, it's 1234 microseconds
3. for `.1234567`, it's 123456 microseconds

Reply:
Do we have a simple rule? The rule for interval is pretty simple: adding `0` at the end until the second fraction has 9 digits, then parse the 9 digits to nanoseconds.

Reply:
> Do we have a simple rule?

I haven't found simpler approach for now. The difference between interval and timestamp is the former one may have time zone at the end or anything else. We cannot say to users don't use the pattern like `mm:ss.SSSSSSXXX yyyy/MM/dd`

Reply:
So what we get here is the milliseconds `FastDateFormat` extracts from the string. I believe `FastDateFormat` can handle the part after seconds. e.g. `12:12:12.1234Z`, the milliseconds part should be 1234.

Reply:
> Can we have some comments to explain the behavior? Seems to me it's
> for .1, it's 1000 microseconds

This is 100 * 1000 microsecond but `SimpleDateFormat` and `FastDateFormat` have weird behavior. The example below on 2.4 without my changes:
```scala
scala> val df = Seq("""{"a":"2019-10-14T09:39:07.1Z"}""").toDF
df: org.apache.spark.sql.DataFrame = [value: string]

scala> val res = df.select(from_json('value, schema, Map("timestampFormat" -> "yyyy-MM-dd'T'HH:mm:ss.SXXX")))
res: org.apache.spark.sql.DataFrame = [jsontostructs(value): struct<a: timestamp>]

scala> res.show(false)
+-------------------------+
|jsontostructs(value)     |
+-------------------------+
|[2019-10-14 12:39:07.001]|
+-------------------------+
```
```scala
scala> val res = df.select(from_json('value, schema, Map("timestampFormat" -> "yyyy-MM-dd'T'HH:mm:ss.SSSXXX")))
res: org.apache.spark.sql.DataFrame = [jsontostructs(value): struct<a: timestamp>]

scala> res.show(false)
+-------------------------+
|jsontostructs(value)     |
+-------------------------+
|[2019-10-14 12:39:07.001]|
+-------------------------+
```
So `.1` cannot be parsed correctly only `0.100`:
```scala
scala> val df = Seq("""{"a":"2019-10-14T09:39:07.100Z"}""").toDF
df: org.apache.spark.sql.DataFrame = [value: string]

scala> val res = df.select(from_json('value, schema, Map("timestampFormat" -> "yyyy-MM-dd'T'HH:mm:ss.SSSXXX")))
res: org.apache.spark.sql.DataFrame = [jsontostructs(value): struct<a: timestamp>]

scala> res.show(false)
+-----------------------+
|jsontostructs(value)   |
+-----------------------+
|[2019-10-14 12:39:07.1]|
+-----------------------+
```

Reply:
What I see in the source code of SimpleDateFormat, it just cast the fraction part to int. `.001` and `.01` are the same and equal to `1`.

Reply:
This check fails unfortunately:
```
    check("yyyy-MM-dd'T'HH:mm:ss.SSSSSSXXX",
      "2019-10-14T09:39:07.000010Z", "2019-10-14T09:39:07.000010Z")
```  
```
Expected :1571045947000010
Actual   :1571045947010000
```
We cannot distinguish `.010` and `.000010` because they are both parsed to `10` inside of `SimpleDateFormat`.

I think this approach cannot work. :-(

Reply:
The commit https://github.com/apache/spark/pull/26507/commits/86ac2b2b0a81aa44b9b9a9133149db15a18dc7cc should fix that